### PR TITLE
access node: skipping the call to the execution node if the transacti…

### PR DESCRIPTION
**Pull request by vishalchangrani**
…on has not been finalized in a GetTransactionResult call

~~An optimization to skip the call to the execution node for the `GetTransactionResult` API call if the transaction has not yet been finalized.~~

This optimization is not needed. When deriving the transaction status a call to `lookupBlock` is made - https://github.com/onflow/flow-go/blob/51f5293513991d6e4319711901e33bab4f4be4a8/engine/access/rpc/backend/backend_transactions.go#L351

The lookupBlock will return an error if the block has not been finalized since the collection for the block would be missing in that case - https://github.com/onflow/flow-go/blob/51f5293513991d6e4319711901e33bab4f4be4a8/engine/access/rpc/backend/backend_transactions.go#L332

Hence, a call to the execution node would anyway be avoided.

Realized this while creating the test case. I am retaining the test case though.

### CODE REVIEW

File: engine/access/rpc/backend/backend_transactions.go [hunk 1]
```diff
@@ -220,14 +220,37 @@ func (b *backendTransactions) GetTransactionResult(
...
+	if status < flow.TransactionStatusFinalized {
+		return &access.TransactionResult{
+			Status:     status,
+			StatusCode: uint(status),
+		}, nil
+	}
+
 	// get events for the transaction
-	executed, events, statusCode, txError, err := b.lookupTransactionResult(ctx, txID)
+	executed, events, _, txError, err := b.lookupTransactionResult(ctx, txID)
```
Comment (vishalchangrani) [inline engine/access/rpc/backend/backend_transactions.go:239]: this call goes upstream to an Execution node.

File: engine/access/rpc/backend/backend_transactions.go [hunk 2]
```diff
@@ -220,14 +220,37 @@ func (b *backendTransactions) GetTransactionResult(
 		return nil, txErr
 	}
 
+	// derive status of the transaction once, assuming that it has not yet executed
```
Comment (vishalchangrani) [inline engine/access/rpc/backend/backend_transactions.go:223]: effecitvely call the `deriveTransactionStatus` twice - once with `executed=false` once with `true`

File: engine/access/rpc/backend/backend_transactions.go [hunk 3]
```diff
@@ -220,14 +220,37 @@ func (b *backendTransactions) GetTransactionResult(
...
+	status, err := b.deriveTransactionStatus(tx, false)
+	if err != nil {
+		return nil, convertStorageError(err)
+	}
+
+	// return the status if it is anything less than finalized (unknown, pending or expired)
+	// since if the transaction is still not finalized, then it would definitely not be sealed
+	if status < flow.TransactionStatusFinalized {
+		return &access.TransactionResult{
```
Comment (vishalchangrani) [inline engine/access/rpc/backend/backend_transactions.go:232]: avoid the call to EN and just return

**Review by jordanschalm — APPROVED** [review]

**Review by Kay-Zee — APPROVED** [review]
Looks good, but any tests for this?

File: engine/access/rpc/backend/backend_test.go [hunk 4]
```diff
@@ -210,9 +210,9 @@ func (suite *Suite) TestGetCollection() {
 	suite.assertAllExpectations()
 }
 
-// TestTransactionStatusTransition tests that the status of transaction changes from Finalized to Sealed
+// TestTransactionFinalizedToSealedStatusTransition tests that the status of transaction changes from Finalized to Sealed
```
Comment (vishalchangrani) [inline engine/access/rpc/backend/backend_test.go:213]: `TestTransactionStatusTransition` renamed to `TestTransactionFinalizedToSealedStatusTransition`
`TestTransactionExpiredStatusTransition` renamed to `TestTransactionPendingToExpiredStatusTransition `
Added new test `TestTransactionPendingToFinalizedStatusTransition`

**Pull request merged**